flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorITCase.java

```diff
@@ -111,6 +106,40 @@ public void testNewTaskExecutorJoinsCluster() throws Exception {
 		miniCluster.requestJobResult(jobGraph.getJobID()).get();
 	}
 
+	/**
+	 * Tests that the job can recover from a failing {@link TaskExecutor}.
+	 */
+	@Test
+	public void testJobRecoveryWithFailingTaskExecutor() throws Exception {
+		final JobGraph jobGraph = createJobGraphWithRestartStrategy(PARALLELISM);
+
+		final CompletableFuture<JobResult> jobResultFuture = submitJobAndWaitUntilRunning(jobGraph);
+
+		// start an additional TaskExecutor
+		miniCluster.startTaskExecutor();
+
+		miniCluster.terminateTaskExecutor(0).get(); // this should fail the job
+
+		BlockingOperator.unblock();
+
+		assertThat(jobResultFuture.get().isSuccess(), is(true));
+	}
+
+	private CompletableFuture<JobResult> submitJobAndWaitUntilRunning(JobGraph jobGraph) throws Exception {
+		miniCluster.submitJob(jobGraph).get();
+
+		final CompletableFuture<JobResult> jobResultFuture = miniCluster.requestJobResult(jobGraph.getJobID());
+
+		assertThat(jobResultFuture.isDone(), is(false));
+
+		CommonTestUtils.waitUntilCondition(
+			jobIsRunning(() -> miniCluster.getExecutionGraph(jobGraph.getJobID())),
+			Deadline.fromNow(TESTING_TIMEOUT),
+			20L);
```

Comment:
There is a bug in `waitUntilCondition` which leads to `20` being ignored.

Reply:
Good catch. Will fix it while merging this PR.